Migrate ExecuTorch's use of pt2e from torch.ao to torchao

metascroy opened the pull request:
Most code related to PT2E quantization is migrating from torch.ao.quantization to torchao.quantization.pt2e.

torchao.quantization.pt2e contains an exact copy of PT2E code in torch.ao.quantization. 

The torchao pin in ExecuTorch has already been bumped pick up these changes.

jerryzh168 commented on backends/apple/coreml/test/test_coreml_quantizer.py:
```diff
@@ -15,12 +15,12 @@
 
 from executorch.backends.apple.coreml.quantizer import CoreMLQuantizer
-from torch.ao.quantization.quantize_pt2e import (
+from torch.export import export_for_training
```
nit: let's use `export`, I heard that's what recommended long term

metascroy replied:
I'd rather not do any changes beyond moving to torchao in this PR.  This existed previously, I think the change is coming from the linter reordering imports.

jerryzh168 replied:
OK, this can be done separately

digantdesai commented on backends/arm/quantizer/arm_quantizer.py:
```diff
@@ -34,25 +34,25 @@
 from executorch.exir.backend.compile_spec_schema import CompileSpec
-from torch.ao.quantization.fake_quantize import (
+from torch.fx import GraphModule, Node
+from torchao.quantization.pt2e import _ObserverOrFakeQuantizeConstructor
```
Correct me if I am wrong but `torchao` isn't a mandetory dep today but now it is?

metascroy replied:
How do we define mandatory dependencies?  It is installed by the install_requirements script?

digantdesai replied:
Seems like we pull in source - https://github.com/pytorch/executorch/blob/647e1f153f13a506a89dfba31eb5d13e82f3a4e0/.gitmodules#L60

So this submodule is already updated since the tests are passing here.

check (1) if we run tests on et wheels with something quant, (2) if we do are they passing for this diff.

digantdesai commented on backends/xnnpack/quantizer/xnnpack_quantizer.py:
```diff
@@ -16,25 +16,25 @@
     QuantizationConfig,
 )
-from torch.ao.quantization.fake_quantize import (
+from torchao.quantization.pt2e.fake_quantize import (
```
nit: there are some affine quantization stuff which assumes torchao isn't always available, can we fix that with this? [here](https://github.com/pytorch/executorch/blob/96c10bb34f349748201905ef44d62da46fee2612/backends/xnnpack/partition/config/quant_affine_configs.py#L36-L41) is one example.

cc @mcr229

cccclai commented:
Adding partners for visibility

digantdesai commented on backends/nxp/quantizer/neutron_quantizer.py:
```diff
@@ -35,9 +35,9 @@
-from torch.ao.quantization.observer import HistogramObserver, MinMaxObserver
-from torch.ao.quantization.quantizer import DerivedQuantizationSpec, Quantizer
-from torch.ao.quantization.quantizer.composable_quantizer import ComposableQuantizer
+from torchao.quantization.pt2e.observer import HistogramObserver, MinMaxObserver
```
cc @robert-kalmar FYI

digantdesai commented on backends/cadence/aot/compiler.py:
```diff
@@ -37,10 +37,10 @@
 
 from torch.export import export
 from torch.export.exported_program import ExportedProgram
+from torchao.quantization.pt2e.quantize_pt2e import convert_pt2e, prepare_pt2e
```
FYI @mcremon-meta

digantdesai commented on backends/arm/quantizer/arm_quantizer.py:
```diff
@@ -34,25 +34,25 @@
 from executorch.exir.backend.compile_spec_schema import CompileSpec
-from torch.ao.quantization.fake_quantize import (
+from torch.fx import GraphModule, Node
+from torchao.quantization.pt2e import _ObserverOrFakeQuantizeConstructor
```
FYI @freddan80 @per

jerryzh168 commented on .lintrunner.toml:
```diff
@@ -371,3 +371,26 @@ command = [
+    '--linter-name=TorchAOImport',
+    '--error-name=Prohibited torch.ao.quantization import',
+    """--error-description=\
+        Imports from torch.ao.quantization are not allowed. \
+        Please import from torchao.quantization.pt2e instead.\
```
also `torch.ao.quantization.quantizer` to `torchao.quantization.quantizer` I think

and optionally can add one more for xnnpack_quantizer for people to import from xnnpack instead of pytorch as well

jerryzh168 replied:
we mainly have 3:

```
# includes all the utils, including observers, fake quants etc.
torchao.quantization.pt2e
# quantizer related objects and utils
torchao.quantization.pt2e.quantizer
# prepare_pt2e, prepare_qat_pt2e, convert_pt2e
torchao.quantization.pt2e.quantize_pt2e

```

metascroy replied:
I will add more lines to the error message to include these.  It should catch xnnpack quantizer from torch.ao.quantization.quantizer already because it throws error on any import from torch.ao.quantization.

metascroy commented:
@cccclai can you have a look at the qualcomm changes?

jerryzh168 commented on backends/example/example_operators/utils.py:
```diff
@@ -4,7 +4,7 @@
 # LICENSE file in the root directory of this source tree.
 
-from torch.ao.quantization.quantizer.quantizer import QuantizationAnnotation
+from torchao.quantization.pt2e.quantizer.quantizer import QuantizationAnnotation
```
single quantizer

jerryzh168 commented on backends/example/example_quantizer.py:
```diff
@@ -11,9 +11,9 @@
-from torch.ao.quantization.pt2e.graph_utils import find_sequential_partitions
-from torch.ao.quantization.quantizer import QuantizationSpec, Quantizer
+from torchao.quantization.pt2e import find_sequential_partitions
+from torchao.quantization.pt2e.observer import HistogramObserver, MinMaxObserver
```
we can remove `observer` here

jerryzh168 commented on backends/mediatek/quantizer/qconfig.py:
```diff
@@ -10,9 +10,9 @@
-from torch.ao.quantization.fake_quantize import FakeQuantize
-from torch.ao.quantization.observer import MinMaxObserver, PerChannelMinMaxObserver
-from torch.ao.quantization.quantizer import QuantizationSpec
+from torchao.quantization.pt2e.fake_quantize import FakeQuantize
+from torchao.quantization.pt2e.observer import MinMaxObserver, PerChannelMinMaxObserver
```
can remove observer and fake quant

metascroy commented on backends/mediatek/quantizer/annotator.py:
```diff
@@ -10,18 +10,18 @@
 
+from torchao.quantization.pt2e.quantizer import (
+    annotate_input_qspec_map,
+    annotate_output_qspec as _annotate_output_qspec,
```
This file uses annotate_output_qspec as the name of some other function they define, so that's why I import it as _annotate_output_qspec.

cccclai commented:
> @cccclai can you have a look at the qualcomm changes?

Looks reasonable to me. Let's just use trunk to trigger more CI jobs.

metascroy commented:
> > @cccclai can you have a look at the qualcomm changes?
> 
> Looks reasonable to me. Let's just use trunk to trigger more CI jobs.

trunk is already triggered.  Thanks!

SS-JIA reviewed (approved):
Vulkan changes LGTM!

metascroy committed:
Migrate ExecuTorch's use of pt2e from torch.ao to torchao (#10294)

Summary:
Most code related to PT2E quantization is migrating from torch.ao.quantization to torchao.quantization.pt2e.

torchao.quantization.pt2e contains an exact copy of PT2E code in torch.ao.quantization.

The torchao pin in ExecuTorch has already been bumped pick up these changes.

Pull Request resolved: https://github.com/pytorch/executorch/pull/10294

Reviewed By: SS-JIA

Differential Revision: D74694311

Pulled By: metascroy

metascroy commented:
Closing this.  The migration was instead done in pieces.
